kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyBackendHandler.java

```diff
@@ -5,99 +5,225 @@
         super.channelReadComplete(ctx);
-        final Channel inboundChannel = inboundCtx.channel();
-        if (unflushedWrites) {
-            unflushedWrites = false;
-            inboundChannel.flush();
+        proxyChannelStateMachine.serverReadComplete();
+    }
+
+    /**
+     * Called by the {@link ProxyChannelStateMachine} to propagate an RPC to the upstream node.
+     * @param msg the RPC to forward.
+     */
+    public void forwardToServer(Object msg) {
+        if (serverCtx == null) {
+            // TODO this shouldn't really be possible to reach. Delete in a releases time once we have more confidence in the StateMachine
+            proxyChannelStateMachine.illegalState("write without outbound active outbound channel");
         }
-        if (!inboundChannel.isWritable()) {
-            ctx.channel().config().setAutoRead(false);
-            this.blockedOutboundCtx = ctx;
+        else {
+            final Channel outboundChannel = serverCtx.channel();
+            if (outboundChannel.isWritable()) {
+                outboundChannel.write(msg, serverCtx.voidPromise());
+                pendingServerFlushes = true;
+            }
+            else {
+                outboundChannel.writeAndFlush(msg, serverCtx.voidPromise());
+                pendingServerFlushes = false;
+            }
         }
+        LOGGER.trace("/READ");
     }
 
-    @Override
-    public void channelInactive(ChannelHandlerContext ctx) {
-        frontendHandler.closeNoResponse(inboundCtx.channel());
+    /**
+     * Called by the {@link ProxyChannelStateMachine} when the bach from the downstream/client side is complete.
+     */
+    public void flushToServer() {
+        if (serverCtx != null) {
+            final Channel serverChannel = serverCtx.channel();
+            if (pendingServerFlushes) {
+                pendingServerFlushes = false;
+                serverChannel.flush();
+            }
+            if (!serverChannel.isWritable()) {
+                proxyChannelStateMachine.onServerUnwritable();
+            }
+        }
     }
 
-    @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        // If the frontEnd has an exception handler for this exception
-        // it's likely to have already dealt with it.
-        // So only act here if its un-expected by the front end.
-        if (exceptionMapper.mapException(cause).isEmpty()) {
-            LOGGER.atWarn()
-                    .setCause(LOGGER.isDebugEnabled() ? cause : null)
-                    .addArgument(cause != null ? cause.getMessage() : "")
-                    .log("Netty caught exception from the backend: {}. Increase log level to DEBUG for stacktrace");
-            frontendHandler.closeNoResponse(ctx.channel());
+    /**
+     * Callback from the {@link ProxyChannelStateMachine} triggered when it wants to apply backpressure to the <em>upstream/server</em> connection
+     */
+    public void blockServerReads() {
```

Comment:
I wonder if `applyBackpressure` and `relieveBackpressure` might be better names?

I also wonder why we've got the `server` in all these names. Isn't it obvious by the fact that this is the backend handler?

Reply:
> I wonder if applyBackpressure and relieveBackpressure might be better names?
Yes, much better names.

>I also wonder why we've got the server in all these names. Isn't it obvious by the fact that this is the backend handler?
I think thats mostly just a fact of how we clarified our understanding and refactored along the way.

Reply:
resolved in 03f955413936f5e37179b555fb84c350ee514f69